### PR TITLE
8273373: Zero: Cannot invoke JVM in primordial threads on Zero

### PR DESCRIPTION
**Pull request by shipilev**
See the bug for more discussion.

Additional testing:
 - [x] Linux Zero x86_64 passes most gtests now (there are unrelated failures)
 - [x] Linux Zero x86_32 passes most gtests now (there are unrelated failures)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273373](https://bugs.openjdk.java.net/browse/JDK-8273373): Zero: Cannot invoke JVM in primordial threads on Zero

### Reviewers
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5376/head:pull/5376` \
`$ git checkout pull/5376`

Update a local copy of the PR: \
`$ git checkout pull/5376` \
`$ git pull https://git.openjdk.java.net/jdk pull/5376/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5376`

View PR using the GUI difftool: \
`$ git pr show -t 5376`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5376.diff">https://git.openjdk.java.net/jdk/pull/5376.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
Hi Aleksey,

The assert in the shared code seems fine. I can't really comment on the Zero changes and the dropping of IA64 support here.

Thanks,
David

File: src/hotspot/os/linux/os_linux.cpp [hunk 1]
```diff
@@ -3366,6 +3366,9 @@ bool os::pd_create_stack_guard_pages(char* addr, size_t size) {
 
     if (mincore((address)stack_extent, os::vm_page_size(), vec) == -1) {
       // Fallback to slow path on all errors, including EAGAIN
+      assert((uintptr_t)addr >= stack_extent,
+             "Sanity: addr should be larger than extent, " PTR_FORMAT " >= " PTR_FORMAT,
+             p2i(addr), stack_extent);
```
Comment (tstuefe) [inline src/hotspot/os/linux/os_linux.cpp:3371]: Should we not do something in non-debug builds too? Maybe guarantee() instead?

**Review by tstuefe — COMMENTED** [review]
Hi,

I do not see a reason why this would be dependent on !zero. Maybe @GoeLin did not want to deal with it, or the problem did not show up on zero in the first place. 

I think you are right in that gtestlauncher is the only launcher where we end up running in the primordial thread.

Removing ia64 code block is fine. Arguably zero may be the last variant where ia64 would still be useful. Debian seems still to build Zero+ia64 in openjdk15: https://buildd.debian.org/status/logs.php?pkg=openjdk-15&arch=ia64

About the zero gtest errors, do you understand why they only happen in other_vm resp. death tests? Something must be special here. I am curious since invoking JNI on the primordial thread may not be that well tested.

I tried to understand the primordial-thread-stack-boundaries coding, but it left a number of questions. For instance we try to find the primordial stack dimensions by:

- read /proc/self/stat, column 28 ("stackstart", bottom address)
- then we go and find the VMA containing this address, and assume the highest address of this VM is the stack top.

I am not sure the latter always works. Is that not dependent on memory layout? It relies on the process stack having its own VMA, but could the kernel not fold it with a neighboring VMA and show it as one line in /proc/self/maps? In that case, the stack start address we calculate could be bogus.

**Comment by tstuefe** [comment]
I got curious and analyzed this further. Let's see if I got this correctly.

When we create the VM on the primordial thread, we locate the stack boundaries for the primordial thread in `os::Linux::capture_initial_stack()`: 

There, we basically look for the boundaries of the primordial thread stack in `/proc/pid/maps`. We manage to locate the correct VMA for the primordial stack. But then we do not simply take the boundary of that VMA as stack boundary. Instead, we take the high VMA address, subtract an assumed stack size (calculated from -Xss and RLIMIT_STACK), then take that address as stack bottom (`_initial_thread_stack_bottom`). 

Seems that that assumed stack size can be smaller than the real size of the stack, resulting in an offset between the real stack VMA low address and `_initial_thread_stack_bottom`. 

That in itself would be not a problem, but in `StackOverflow::create_stack_guard_pages()` we place the guard pages at an address which we calculate from `os::current_stack_base()`. `os::current_stack_base()` calls `current_stack_region()`, which on Zero just calls `pthread_attr_getstack(3)`, ignoring the previously captured `_initial_thread_stack_bottom`.

In `os::pd_create_stack_guard_pages()` we then essentially use both values -  `os::current_stack_base()` and `_initial_thread_stack_bottom` - to calculate the arguments to `get_stack_commited_bottom()`. The resulting assert is caused by mixing those conflicting values.

Bottomline, we should either be using `pthread_attr_getstack(3)` or our `_initial_thread_stack_(bottom|size)` but not mix them since they may differ. Therefore the proposed patch is correct. The non-zero version of `current_stack_region()` in os_linux.cpp handles the primordial thread stack correctly.

---

There are a number of additional beauty spots, like the fact that in `get_stack_commited_bottom()` the `pages` variable is 32bit unsigned. So the assert itself is random in that it depends on whether or not the overflow produces a negative overflow.

Also, I don't understand the logic in `os::pd_create_stack_guard_pages()` where the "fallback" is to call mincore() with what basically would be a zero size? Since both `addr` and `stack_extend` are the stack bottom.

---

I would probably rename the issue to "Cannot invoke JVM in primordial threads on Zero".

Cheers, Thomas

File: src/hotspot/os/linux/os_linux.cpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/os/linux/os_linux.cpp:3371]: Nah, I think `assert` is okay here. It is supposed to aid debugging, not crash hard.

**Comment by shipilev** [comment]
> Bottomline, we should either be using `pthread_attr_getstack(3)` or our `_initial_thread_stack_(bottom|size)` but not mix them since they may differ. Therefore the proposed patch is correct. The non-zero version of `current_stack_region()` in os_linux.cpp handles the primordial thread stack correctly.

Yes, that seems to be a gist of it. In fact, if I reply `_initial_thread_stack_(bottom|size)` in Zero version of `current_stack_region()` for a primordial thread, gtests also start to pass. But I gather there is no need to continue supporting Zero-specific `current_stack_region()`, and thus dodge the issue altogether by using the code the rest of VM code uses.

**Comment by shipilev** [comment]
I looked around Debian's support for IA64, and I think we don't want to drop the support for it just yet. Instead of adding the `IA64` block to a shared code, I shall try to make a more conservative fix to Zero-specific code then...

**Comment by tstuefe** [comment]
> > Bottomline, we should either be using `pthread_attr_getstack(3)` or our `_initial_thread_stack_(bottom|size)` but not mix them since they may differ. Therefore the proposed patch is correct. The non-zero version of `current_stack_region()` in os_linux.cpp handles the primordial thread stack correctly.
> 
> Yes, that seems to be the gist of it. In fact, if I reply `_initial_thread_stack_(bottom|size)` in Zero version of `current_stack_region()` for a primordial thread, gtests also start to pass. But I gather there is no need to continue supporting Zero-specific `current_stack_region()`, and thus dodge the issue altogether by using the code the rest of VM code uses.

Yes, and zero also needs Goetz' fix for the glibc guard size problem in JDK-8169373.

**Comment by tstuefe** [comment]
> I looked around Debian's support for IA64, and I think we don't want to drop the support for it just yet. Zero VM is one of those last stands for weird arches portability :) Instead of adding the `IA64` block to a shared code, I shall try to make a more conservative fix to Zero-specific code then...

I'd be surprised if it still works in JDK 18 though. The latest they build is 15. ia64 breaks so easily, and no one took care of it in the recent past.

**Comment by shipilev** [comment]
> I'd be surprised if it still works in JDK 18 though. The latest they build is 15. ia64 breaks so easily, and no one took care of it in the recent past.

Yes, but I also want the fix to be backportable to JDK 11. Which means I should probably keep whatever IA64 support there is. I'll figure this out.

**Comment by tstuefe** [comment]
> > I'd be surprised if it still works in JDK 18 though. The latest they build is 15. ia64 breaks so easily, and no one took care of it in the recent past.
> 
> Yes, but I also want the fix to be backportable to JDK 11. Which means I should probably keep whatever IA64 support there is. I'll figure this out.

Oh right. Makes sense.

**Comment by shipilev** [comment]
I pushed the most conservative version I could think of now. Debian folks seem to also patch this method for HPPA: it seems to have unusual stack arrangements too, so I left the original method body as untouched as possible. This still makes gtests happier for Zero.

**Review by tstuefe — APPROVED** [review]
Looks good

**Comment by shipilev** [comment]
Thank you!

/integrate

**Pull request closed without merging**